src/workerd/api/node/zlib-util.c++

```diff
@@ -450,4 +547,67 @@ void ZlibUtil::ZlibStream::reset(jsg::Lock& js) {
   }
 }
 
+kj::Array<kj::byte> syncProcessBuffer(ZlibContext& ctx, GrowableBuffer& result) {
+  KJ_DEFER(ctx.close());
+
+  do {
+    // TODO: We respect the chunk size to match node, but it's a bit silly.. should we just
+    // override it for perf?
+    result.addChunk();
+    ctx.setOutputBuffer(kj::ArrayPtr(result.end(), result.available()));
+
+    ctx.work();
+
+    KJ_IF_SOME(error, ctx.getError()) {
+      JSG_FAIL_REQUIRE(Error, error.message);
+    }
+
+    result.adjustUnused(ctx.getAvailOut());
+  } while (ctx.getAvailOut() == 0);
+
+  KJ_IF_SOME(error, ctx.getError()) {
+    JSG_FAIL_REQUIRE(Error, error.message);
+  }
+
+  return result.releaseAsArray();
+}
+
+kj::Array<kj::byte> ZlibUtil::zlibSync(InputSource data, Options opts, ZlibModeValue mode) {
+  ZlibContext ctx;
+
+  auto chunkSize = opts.chunkSize.orDefault(Z_DEFAULT_CHUNK);
+  auto maxOutputLength = opts.maxOutputLength.orDefault(Z_MAX_CHUNK);
+
+  JSG_REQUIRE(Z_MIN_CHUNK <= chunkSize && chunkSize <= Z_MAX_CHUNK, Error, "Invalid chunkSize");
+  JSG_REQUIRE(maxOutputLength <= Z_MAX_CHUNK, Error, "Invalid maxOutputLength");
+  GrowableBuffer result(chunkSize, maxOutputLength);
+
+  ctx.setMode(static_cast<ZlibMode>(mode));
+  ctx.initialize(opts.level.orDefault(Z_DEFAULT_LEVEL),
+      opts.windowBits.orDefault(Z_DEFAULT_WINDOWBITS), opts.memLevel.orDefault(Z_DEFAULT_MEMLEVEL),
+      opts.strategy.orDefault(Z_DEFAULT_STRATEGY), kj::mv(opts.dictionary));
+  ctx.setFlush(opts.finishFlush.orDefault(Z_FINISH));
+
+  KJ_SWITCH_ONEOF(data) {
+    KJ_CASE_ONEOF(dataBuf, kj::Array<kj::byte>) {
+      ctx.setInputBuffer(dataBuf);
+    }
+
+    KJ_CASE_ONEOF(dataStr, jsg::NonCoercible<kj::String>) {
+      ctx.setInputBuffer(dataStr.value.asBytes());
+    }
+  }
+
+  return syncProcessBuffer(ctx, result);
```

Comment:
nit: may be useful here to add a comment reminding folks that `ctx` will be closed when `syncProcessBuffer` returns.

Reply:
alternatively, would it make sense to have `~ZlibContext` call `close()` if necessary so that we don't have to worry about the `KJ_DEFER(ctx.close())` call?

Reply:
I wish ZlibContext was RAII so I didn't have to do this manually but maybe it has to be that way to match NodeJS? What do you think @anonrig?

Reply:
Heh James and I commented the same thing at the same time

Reply:
The context shouldn't live without the CompressionStream class. Since this change diverges the implementation of one-shot classes from Node.js, I don't think we should diverge more. It is complicated as it is right now.

Reply:
I added a wrapper called ZlibContextRAII. It's a bit more verbose but I find this the more intuitive approach for me.